test/e2e/network/netpol/k8s_util.go

```diff
@@ -0,0 +1,373 @@
+	}
+
+	return createdService, err
+}
+
+// CreateOrUpdateDeployment is a convenience function for idempotent setup of deployments
+func (k *Kubernetes) CreateOrUpdateDeployment(deployment *appsv1.Deployment) (*appsv1.Deployment, error) {
+	ns := deployment.Namespace
+	log.Infof("creating/updating deployment %s/%s", ns, deployment.Name)
+
+	createdDeployment, err := k.ClientSet.AppsV1().Deployments(ns).Create(context.TODO(), deployment, metav1.CreateOptions{})
+	if err == nil {
+		log.Infof("created deployment %s/%s", ns, createdDeployment.Name)
+		return createdDeployment, nil
+	}
+
+	log.Debugf("unable to create deployment %s/%s, let's try update instead", ns, deployment.Name)
+	createdDeployment, err = k.ClientSet.AppsV1().Deployments(ns).Update(context.TODO(), createdDeployment, metav1.UpdateOptions{})
+	if err != nil {
+		log.Errorf("unable to create/update deployment %s/%s: %+v", ns, deployment.Name, err)
+	}
+
+	return createdDeployment, errors.Wrapf(err, "unable to update deployment %s/%s", ns, deployment.Name)
+}
+
+// CleanNetworkPolicies is a convenience function for deleting network policies before startup of any new test.
+func (k *Kubernetes) CleanNetworkPolicies(namespaces []string) error {
+	for _, ns := range namespaces {
+		log.Infof("deleting policies..........%v ", ns)
+		l, err := k.ClientSet.NetworkingV1().NetworkPolicies(ns).List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "unable to list network policies in ns %s", ns)
+		}
+		for _, np := range l.Items {
+			log.Infof("deleting network policy %s in ns %s", np.Name, ns)
+			err = k.ClientSet.NetworkingV1().NetworkPolicies(ns).Delete(context.TODO(), np.Name, metav1.DeleteOptions{})
+			if err != nil {
+				return errors.Wrapf(err, "unable to delete network policy %s", np.Name)
+			}
+		}
+	}
+	return nil
+}
+
+// ClearCache clears the kube pod cache
+func (k *Kubernetes) ClearCache() {
+	log.Info("Clearing pod cache")
+	k.mutex.Lock()
+	k.podCache = map[string][]v1.Pod{}
+	k.mutex.Unlock()
+	log.Info("Pod cache successfully cleared")
+}
+
+// CreateOrUpdateNetworkPolicy is a convenience function for updating/creating netpols
+func (k *Kubernetes) CreateOrUpdateNetworkPolicy(ns string, netpol *networkingv1.NetworkPolicy) (*networkingv1.NetworkPolicy, error) {
+	log.Infof("creating/updating network policy %s in ns %s", netpol.Name, ns)
+	netpol.ObjectMeta.Namespace = ns
+	np, err := k.ClientSet.NetworkingV1().NetworkPolicies(ns).Update(context.TODO(), netpol, metav1.UpdateOptions{})
+	if err == nil {
+		return np, err
+	}
+
+	log.Debugf("unable to update network policy %s in ns %s, let's try creating it instead (error: %s)", netpol.Name, ns, err)
+	np, err = k.ClientSet.NetworkingV1().NetworkPolicies(ns).Create(context.TODO(), netpol, metav1.CreateOptions{})
+	if err != nil {
+		log.Debugf("unable to create network policy: %s", err)
+	}
+	return np, err
+}
+
+func (k *Kubernetes) waitForHTTPServers(model *Model) error {
```

Comment:
you can wait for the deployment to be complete
https://github.com/kubernetes/kubernetes/blob/1a66eb7b8aba8d5e66d855d6b8b31fdeab648778/test/e2e/apps/deployment.go#L256

Reply:
👍

Reply:
hmm, so I thought in this function the idea was to actually specifically  probe all the servers before starting tests...  not just wait for the containers to be up and running  ?

Reply:
added a comment to clarify that

Reply:
another option is to set readiness probe and wait until the pod is running

Reply:
true but, I think running the probe here is a verification that end to end probing is working + pod is up though, so we'd be losing that extra verification we moved to a readiness check.